File: scripts/lib/verification.ts
import { readFile } from 'node:fs/promises';
import { log } from './log.js';

export interface VerificationRecord {
  source: string;
  lastVerified: string;
}

const ISO_DATE_RE = /^\d{4}-\d{2}-\d{2}$/;

/**
 * Parse the `## Verification` section of DATA_SOURCES.md and extract the
 * `Last verified: YYYY-MM-DD` line for each `### <source>` subsection.
 * Sources whose subsection is missing or whose date is malformed are
 * logged at warn level and omitted from the result.
 */
export function parseVerificationMarkdown(markdown: string): VerificationRecord[] {
  const start = markdown.indexOf('\n## Verification');
  if (start === -1) {
    log.warn('verification: no "## Verification" section found in markdown');
    return [];
  }

  const tail = markdown.slice(start + 1);
  const nextTopLevel = tail.search(/\n## (?!#)/);
  const section = nextTopLevel === -1 ? tail : tail.slice(0, nextTopLevel);

  const records: VerificationRecord[] = [];
  const subsectionRe = /\n### (\S+)\n([\s\S]*?)(?=\n### |\n## |$)/g;
  let match: RegExpExecArray | null;
  while ((match = subsectionRe.exec(section)) !== null) {
    const source = (match[1] ?? '').trim();
    const body = match[2] ?? '';
    if (!source) continue;
    const dateMatch = body.match(/Last verified:\s*(\S+)/i);
    if (!dateMatch || !dateMatch[1]) {
      log.warn({ source }, 'verification: subsection has no "Last verified" line');
      continue;
    }
    const date = dateMatch[1].trim();
    if (!ISO_DATE_RE.test(date)) {
      log.warn({ source, date }, 'verification: malformed date; expected YYYY-MM-DD');
      continue;
    }
    records.push({ source, lastVerified: date });
  }
  return records;
}

export async function readVerificationFromMarkdown(path: string): Promise<VerificationRecord[]> {
  try {
    const text = await readFile(path, 'utf8');
    return parseVerificationMarkdown(text);
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    log.warn({ path, err: msg }, 'verification: could not read markdown file');
    return [];
  }
}

File: go/internal/transform/verification.go
package transform

import (
	"os"
	"regexp"
	"strings"
)

// VerificationRecord pairs a source name with its last-verified ISO date.
type VerificationRecord struct {
	Source       string
	LastVerified string
}

var (
	isoDateRe     = regexp.MustCompile(`^\d{4}-\d{2}-\d{2}$`)
	subsectionRe  = regexp.MustCompile(`(?m)^### (\S+)$`)
	lastVerifyRe  = regexp.MustCompile(`(?i)Last verified:\s*(\S+)`)
)

// ParseVerificationMarkdown extracts "Last verified: YYYY-MM-DD" entries from
// the `## Verification` section of DATA_SOURCES.md, keyed by `### <source>`.
// Mirrors scripts/lib/verification.ts.
func ParseVerificationMarkdown(markdown string) []VerificationRecord {
	const header = "\n## Verification"
	start := strings.Index(markdown, header)
	if start == -1 {
		return nil
	}
	tail := markdown[start+1:]
	// Section ends at the next top-level "## " heading (but not "### ").
	end := len(tail)
	// search for "\n## " not "\n###"
	for i := 0; i < len(tail)-3; i++ {
		if tail[i] == '\n' && tail[i+1] == '#' && tail[i+2] == '#' && tail[i+3] == ' ' {
			// require char i+4 != '#' (which would have been matched as "### " — but since we required " ", we're good)
			if i > 0 { // skip the initial "## Verification" header at i==0 after slicing
				end = i
				break
			}
		}
	}
	section := tail[:end]

	// Find each "### <source>" subsection.
	loc := subsectionRe.FindAllStringSubmatchIndex(section, -1)
	if loc == nil {
		return nil
	}

	var out []VerificationRecord
	for i, m := range loc {
		nameStart, nameEnd := m[2], m[3]
		source := strings.TrimSpace(section[nameStart:nameEnd])
		if source == "" {
			continue
		}
		bodyStart := m[1]
		bodyEnd := len(section)
		if i+1 < len(loc) {
			bodyEnd = loc[i+1][0]
		}
		body := section[bodyStart:bodyEnd]
		dm := lastVerifyRe.FindStringSubmatch(body)
		if len(dm) < 2 {
			continue
		}
		date := strings.TrimSpace(dm[1])
		if !isoDateRe.MatchString(date) {
			continue
		}
		out = append(out, VerificationRecord{Source: source, LastVerified: date})
	}
	return out
}

// ReadVerificationFromMarkdown reads a DATA_SOURCES.md-style file and parses
// it. Missing or unreadable files yield an empty result rather than an error.
func ReadVerificationFromMarkdown(path string) []VerificationRecord {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil
	}
	return ParseVerificationMarkdown(string(data))
}
